common/beerocks/bcl/source/son/son_wireless_utils.cpp

```diff
@@ -55,43 +56,110 @@ wireless_utils::estimate_ul_params(int ul_rssi, uint16_t sta_phy_tx_rate_100kb,
+               << " | max_ant_mode:" << max_ant_mode << " | max MCS:" << max_mcs << " | max BW:"
+               << beerocks::utils::convert_bandwidth_to_int((beerocks::eWiFiBandwidth)max_bw);
+
+    if (ul_rssi == beerocks::RSSI_INVALID) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (invalid RSSI)";
+        return estimation_invalid;
+    }
+
+    // If station phyrate value is below table's minimum, return minimal estimation
+    if (sta_phy_tx_rate_100kb < phy_rate_table[0][0].bw_values[0].gi_long_rate) {
+        LOG(DEBUG) << __FUNCTION__ << " | Can not estimate UL parameters (STA phyrate is too low)";
+        return estimation_minimal;
+    }
+
+    // If station phyrate value is above table's maximum, return maximal estimation
+    if (sta_phy_tx_rate_100kb >
+        phy_rate_table[max_ant_mode][max_mcs].bw_values[max_bw].gi_short_rate) {
+        LOG(DEBUG) << "STA phy rate (" << sta_phy_tx_rate_100kb / 10
+                   << " Mbps) is above maximum possible in current MCS/NSS/BW mode ("
+                   << phy_rate_table[max_ant_mode][max_mcs].bw_values[max_bw].gi_short_rate / 10
+                   << " Mbps)";
+
+        estimation_normal.status   = ESTIMATION_SUCCESS;
+        estimation_normal.tx_power = is_5ghz ? phy_rate_table[max_ant_mode][max_mcs].tx_power_5
+                                             : phy_rate_table[max_ant_mode][max_mcs].tx_power_2_4;
+        estimation_normal.rssi =
+            int(ceil(phy_rate_table[max_ant_mode][max_mcs].bw_values[max_bw].rssi / 10.0));
+
+        LOG(DEBUG) << "Return maximal estimation values | tx_power:" << estimation_normal.tx_power
+                   << " | RSSI:" << estimation_normal.rssi;
+
+        return estimation_normal;
+    }
+
+    for (int ant_mode = max_ant_mode; ant_mode > -1; ant_mode--) { // filter by ant_mode
+        for (int bw = max_bw; bw > -1; bw--) {                     // filter by max_bw
             for (int mcs = max_mcs; mcs > -1; mcs--) {             // filter by mcs
 
-                estimated_ul_rssi_lut = phy_rate_table[ant_mode][mcs].bw_values[bw].rssi;
-                estimated_ul_rssi_lut_delta =
-                    (int)std::abs((float)(ul_rssi_lut - estimated_ul_rssi_lut));
+                estimated_ul_rssi_lut       = phy_rate_table[ant_mode][mcs].bw_values[bw].rssi;
+                estimated_ul_rssi_lut_delta = std::abs(ul_rssi_lut - estimated_ul_rssi_lut);
 
-                // same rate && min rssi_delta
-                if ((sta_phy_tx_rate_100kb ==
-                     phy_rate_table[ant_mode][mcs].bw_values[bw].gi_long_rate) ||
-                    (sta_phy_tx_rate_100kb ==
-                     phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate)) {
+                auto gi_long_rate  = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_long_rate;
+                auto gi_short_rate = phy_rate_table[ant_mode][mcs].bw_values[bw].gi_short_rate;
```

Comment:
how it will make it shorter? We will have to write `value.gi_long_rate` and  `value.gi_short_rate` instead of `gi_long_rate` and `gi_short_rate`.
Anyway, it doesn't seem critical. I leaving it as is.